Add a timeout argument to RPC shutdown()

dracifer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65425



cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @gcramer23

Differential Revision: [D31092483](https://our.internmc.facebook.com/intern/diff/D31092483)

dracifer committed:
Add a timeout argument to RPC shutdown()

[ghstack-poisoned]

dracifer committed:
Update on "Add a timeout argument to RPC shutdown()"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31092483](https://our.internmc.facebook.com/intern/diff/D31092483)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1319,6 +1319,35 @@ def _wait_all_workers(self, f, x):
+
+        og_func = rpc.api._wait_all_workers
+
+        def wait_all_workers_sleep(timeout):
```
This is a pretty clever way to test it! I was originally thinking about sth like letting rank0 using an self-RPC to occupy its thread to trigger the timeout. I think yours is better, and less depend on the RPC threading model.

mrshenli reviewed (approved):
LGTM!

dracifer committed:
Update on "Add a timeout argument to RPC shutdown()"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

Differential Revision: [D31092483](https://our.internmc.facebook.com/intern/diff/D31092483)

[ghstack-poisoned]

dracifer commented:
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31092483).
